azure/scope/machinepool.go

```diff
@@ -153,61 +156,291 @@ func (m *MachinePoolScope) ProvisioningState() infrav1.VMState {
+func (m *MachinePoolScope) NeedsRequeue() bool {
+	if m.vmssState != nil {
+		for _, machine := range m.vmssState.Instances {
+			if !machine.LatestModelApplied {
+				return true
+			}
+		}
+	}
+
+	state := m.AzureMachinePool.Status.ProvisioningState
+	return state != nil && !infrav1.IsTerminalVMState(*state)
+}
+
+// updateReplicasAndProviderIDs ties the Azure VMSS instance data and the Node status data together to build and update
+// the AzureMachinePool replica count and providerIDList.
+func (m *MachinePoolScope) updateReplicasAndProviderIDs(ctx context.Context, instances []infrav1exp.VMSSVM) error {
 	ctx, span := tele.Tracer().Start(ctx, "scope.MachinePoolScope.UpdateInstanceStatuses")
 	defer span.End()
 
-	providerIDs := make([]string, len(instances))
-	for i, instance := range instances {
-		providerIDs[i] = fmt.Sprintf("azure://%s", instance.ID)
-	}
-
-	nodeStatusByProviderID, err := m.getNodeStatusByProviderID(ctx, providerIDs)
+	machines, err := m.getMachinePoolMachines(ctx)
 	if err != nil {
-		return errors.Wrap(err, "failed to get node status by provider id")
+		return errors.Wrap(err, "failed to get machine pool machines")
 	}
 
 	var readyReplicas int32
-	instanceStatuses := make([]*infrav1exp.AzureMachinePoolInstanceStatus, len(instances))
-	for i, instance := range instances {
-		instanceStatuses[i] = &infrav1exp.AzureMachinePoolInstanceStatus{
-			ProviderID:        fmt.Sprintf("azure://%s", instance.ID),
-			InstanceID:        instance.InstanceID,
-			InstanceName:      instance.Name,
-			ProvisioningState: &instance.State,
+	providerIDs := make([]string, len(machines))
+	for i, machine := range machines {
+		if machine.Status.Ready {
+			readyReplicas++
 		}
+		providerIDs[i] = machine.Spec.ProviderID
+	}
 
-		instanceStatus := instanceStatuses[i]
-		if nodeStatus, ok := nodeStatusByProviderID[instanceStatus.ProviderID]; ok {
-			instanceStatus.Version = nodeStatus.Version
-			if m.MachinePool.Spec.Template.Spec.Version != nil {
-				instanceStatus.LatestModelApplied = instanceStatus.Version == *m.MachinePool.Spec.Template.Spec.Version
-			}
+	m.AzureMachinePool.Status.Replicas = readyReplicas
+	m.AzureMachinePool.Spec.ProviderIDList = providerIDs
+	return nil
+}
+
+func (m *MachinePoolScope) getMachinePoolMachines(ctx context.Context) ([]infrav1exp.AzureMachinePoolMachine, error) {
+	ctx, span := tele.Tracer().Start(ctx, "scope.MachinePoolScope.getMachinePoolMachines")
+	defer span.End()
+
+	labels := map[string]string{
+		clusterv1.ClusterLabelName:      m.ClusterName(),
+		infrav1exp.MachinePoolNameLabel: m.AzureMachinePool.Name,
+	}
+	ampml := &infrav1exp.AzureMachinePoolMachineList{}
+	if err := m.client.List(ctx, ampml, client.InNamespace(m.AzureMachinePool.Namespace), client.MatchingLabels(labels)); err != nil {
+		return nil, errors.Wrap(err, "failed to list AzureMachinePoolMachines")
+	}
+
+	return ampml.Items, nil
+}
+
+func (m *MachinePoolScope) applyAzureMachinePoolMachines(ctx context.Context) error {
```

Comment:
This contains the logic to compare the state of Azure VMSS with the state of `AzureMachinePool(Machine)s`. If a machine exists in Azure, a `AzureMachinePoolMachine` is created. If an `AzureMachinePoolMachine` exists, but doesn't have an Azure counterpart, it is asked to delete. At the end, the upgrade / over-provision state is evaluated and machines can be deleted if the state requires.